include/sacn/cpp/source_detector.h

```diff
@@ -0,0 +1,402 @@
+     * platforms), and the configuration you pass to sacn_source_detector_create() has source_count_max and
+     * universes_per_source_max set to #SACN_SOURCE_DETECTOR_INFINITE, this callback will never be called.
+     *
+     * If #SACN_DYNAMIC_MEM was defined to 0 when sACN was compiled, source_count_max and universes_per_source_max are
+     * ignored and #SACN_SOURCE_DETECTOR_MAX_SOURCES and #SACN_SOURCE_DETECTOR_MAX_UNIVERSES_PER_SOURCE are used
+     * instead.
+     *
+     * This callback is rate-limited: it will only be called when the first universe discovery packet is received that
+     * takes the module beyond a memory limit.  After that, it will not be called until the number of sources or
+     * universes has dropped below the limit and hits it again.
+     */
+    virtual void HandleMemoryLimitExceeded() {}
+  };
+
+  /**
+   * @ingroup sacn_source_detector_cpp
+   * @brief A set of configuration settings that a source detector needs to initialize.
+   */
+  struct Settings
+  {
+    /********* Required values **********/
+
+    /********* Optional values **********/
+
+    /** The maximum number of sources this detector will record.  It is recommended that applications using dynamic
+       memory use #SACN_SOURCE_DETECTOR_INFINITE for this value. This parameter is ignored when configured to use
+       static memory -- #SACN_SOURCE_DETECTOR_MAX_SOURCES is used instead.*/
+    int source_count_max{SACN_SOURCE_DETECTOR_INFINITE};
+
+    /** The maximum number of universes this detector will record for a source.  It is recommended that applications
+       using dynamic memory use #SACN_SOURCE_DETECTOR_INFINITE for this value. This parameter is ignored when
+       configured to use static memory -- #SACN_SOURCE_DETECTOR_MAX_UNIVERSES_PER_SOURCE is used instead.*/
+    int universes_per_source_max{SACN_SOURCE_DETECTOR_INFINITE};
+
+    /** Create default data structure. */
+    Settings() = default;
+  };
+
+  SourceDetector() = default;
+  SourceDetector(const SourceDetector& other) = delete;
+  SourceDetector& operator=(const SourceDetector& other) = delete;
+  SourceDetector(SourceDetector&& other) = default;            /**< Move a detector instance. */
+  SourceDetector& operator=(SourceDetector&& other) = default; /**< Move a detector instance. */
+
+  etcpal::Error Startup(const Settings& settings, NotifyHandler& notify_handler,
+                        std::vector<SacnMcastInterface>& netints = kAllInterfaces);
+
+  etcpal::Error Startup(NotifyHandler& notify_handler, std::vector<SacnMcastInterface>& netints = kAllInterfaces);
+  void Shutdown();
+  etcpal::Error ResetNetworking(std::vector<SacnMcastInterface>& netints = kAllInterfaces);
+
+  constexpr Handle handle() const;
+
+private:
+  SacnSourceDetectorConfig TranslateConfig(const Settings& settings, NotifyHandler& notify_handler);
+
+  /** An internal shortcut for selecting all network interfaces. **/
+  static std::vector<SacnMcastInterface> kAllInterfaces;
```

Comment:
Does this really compile without being defined somewhere?

Reply:
You're right -- that's bizarre.  I was focusing on how to make the default parameter work, but this doesn't make sense.  I'm wondering if I should remove this entirely and make the caller pass in the empty vector again.  The difficulty is that this is an in/out parameter, so you can't just have `foo(std::vector<SacnMcastInterface>& param = std::vector<SacnMcastInterface>());`  What do you think, @samkearney?

Reply:
> The difficulty is that this is an in/out parameter, so you can't just have `foo(std::vector<SacnMcastInterface>& param = std::vector<SacnMcastInterface>());`

Wait, why not?

Reply:
Because it won't compile. :)

Reply:
The exact error is:
`3>D:\Projects\sACN\include\sacn/cpp/source_detector.h(206,66): error C2440: 'default argument': cannot convert from 'std::vector<SacnMcastInterface,std::allocator<SacnMcastInterface>>' to 'std::vector<SacnMcastInterface,std::allocator<SacnMcastInterface>> &'
3>D:\Projects\sACN\include\sacn/cpp/source_detector.h(206,66): message : A non-const reference may only be bound to an lvalue
`

Reply:
Oh, I see. I understand why that language limitation exists, although it hurts us in this case.

OK, how about the kAllInterfaces member, but make it non-static?

Reply:
`2>D:\Projects\sACN\include\sacn/cpp/source_detector.h(206,66): error C2648: 'sacn::SourceDetector::kAllInterfaces': use of member as default parameter requires static member
2>D:\Projects\sACN\include\sacn/cpp/source_detector.h(218): message : see declaration of 'sacn::SourceDetector::kAllInterfaces'
2>D:\Projects\sACN\include\sacn/cpp/source_detector.h(206,66): error C2355: 'this': can only be referenced inside non-static member functions or non-static data member initializers
`
Keep the ideas coming!  I thought I tried everything, but you might see something I missed.

Reply:
More overloads of Startup():
```cpp
etcpal::Error Startup(NotifyHandler& notify_handler);
etcpal::Error Startup(const Settings& settings, NotifyHandler& notify_handler);
etcpal::Error Startup(NotifyHandler& notify_handler, std::vector<SacnMcastInterface>& netints);
etcpal::Error Startup(const Settings& settings, NotifyHandler& notify_handler, std::vector<SacnMcastInterface>& netints);
```

Reply:
I was trying to find something more elegant, but I guess this is probably the best bet under the circumstances.  I'll make this change.  Thanks!

Reply:
Ok, my updates should be done.  Let me know if you have anything else.